consensus/tendermint/message_store.go

```diff
@@ -0,0 +1,310 @@
+	valid map[common.Hash]struct{}
+	// messages maps message hash to message.
+	messages map[common.Hash]*message
+	// rawMessages maps message hash to raw message bytes.
+	rawMessages map[common.Hash][]byte
+	// values maps value hash to value.
+	values map[common.Hash]*types.Block
+	// messageBounds keeps track of the heights for which the messageStore retains messages.
+	messageBounds *bounds
+}
+
+func (m *messageStore) heightMessages(height uint64) []*message {
+	var messages []*message
+	for _, msgTypeMap := range m.msgHashes[height] {
+		for _, addressMap := range msgTypeMap {
+			for _, hash := range addressMap {
+				messages = append(messages, m.messages[hash])
+			}
+		}
+	}
+	return messages
+}
+
+func (m *messageStore) rawHeightMessages(height uint64) [][]byte {
+	var messages [][]byte
+	for _, msgTypeMap := range m.msgHashes[height] {
+		for _, addressMap := range msgTypeMap {
+			for _, hash := range addressMap {
+				messages = append(messages, m.rawMessages[hash])
+			}
+		}
+	}
+	return messages
+}
+
+func newMessageStore(messageBounds *bounds) *messageStore {
+	return &messageStore{
+		msgHashes:     make(map[uint64]map[int64]map[algorithm.Step]map[common.Address]common.Hash),
+		rawMessages:   make(map[common.Hash][]byte),
+		messages:      make(map[common.Hash]*message),
+		valid:         make(map[common.Hash]struct{}),
+		values:        make(map[common.Hash]*types.Block),
+		messageBounds: messageBounds,
+	}
+
+}
+
+func (m *messageStore) Message(h common.Hash) *message {
+	return m.messages[h]
+}
+
+func (m *messageStore) signatures(value algorithm.ValueID, round int64, height uint64) [][]byte {
+	var sigs [][]byte
+
+	// println("signatures -----")
+	for _, msgHash := range m.msgHashes[height][round][algorithm.Precommit] {
+		// spew.Dump(m.rawMessages[msgHash].decodedMsg)
+		if value == m.messages[msgHash].consensusMessage.Value {
+			sigs = append(sigs, m.messages[msgHash].signature)
+		}
+	}
+	// println("----------------")
+	return sigs
+}
+
+func (m *messageStore) prevoteQuorum(valueHash *common.Hash, round int64, header *types.Header) bool {
+	msgType := new(algorithm.Step)
+	*msgType = algorithm.Prevote
+	// println("prevote power --------")
+	vp := m.votePower(valueHash, round, msgType, header) >= header.Committee.Quorum()
+	// println(vp, "----------------")
+	return vp
+}
+
+func (m *messageStore) precommitQuorum(valueHash *common.Hash, round int64, header *types.Header) bool {
+	msgType := new(algorithm.Step)
+	*msgType = algorithm.Precommit
+	// // println("precommit power", m.votePower(valueHash, round, msgType, header))
+	// println("precommit power --------")
+	vp := m.votePower(valueHash, round, msgType, header) >= header.Committee.Quorum()
+	// println(vp, "----------------")
+	return vp
+}
+
+func (m *messageStore) fail(round int64, header *types.Header) bool {
+	// println("fail power --------")
+	vp := m.votePower(nil, round, nil, header) >= header.Committee.Quorum()
```

Comment:
Hi guys, hope you're all well!

```suggestion
	vp := m.votePower(nil, round, nil, header) >= header.Committee.F()
```